snuba/datasets/entities/discover.py

```diff
@@ -317,39 +328,83 @@ def __init__(self) -> None:
+                selector=EventsQueryStorageSelector(
+                    mappers=events_translation_mappers.concat(
+                        transaction_translation_mappers
+                    )
+                    .concat(null_function_translation_mappers)
+                    .concat(
+                        TranslationMappers(
+                            # XXX: Remove once we are using errors
+                            columns=[
+                                ColumnToMapping(
+                                    None, "release", None, "tags", "sentry:release"
+                                ),
+                                ColumnToMapping(
+                                    None, "dist", None, "tags", "sentry:dist"
+                                ),
+                                ColumnToMapping(
+                                    None, "user", None, "tags", "sentry:user"
+                                ),
+                            ],
+                            subscriptables=[
+                                SubscriptableMapper(None, "tags", None, "tags"),
+                                SubscriptableMapper(None, "contexts", None, "contexts"),
+                            ],
                         )
                     )
+                )
+            ),
+        )
+
+        discover_storage = get_storage(StorageKey.DISCOVER)
+
+        discover_pipeline_builder = SimplePipelineBuilder(
+            query_plan_builder=SingleStorageQueryPlanBuilder(
+                storage=discover_storage,
+                mappers=events_translation_mappers.concat(
+                    transaction_translation_mappers
+                )
+                .concat(null_function_translation_mappers)
+                .concat(
+                    TranslationMappers(
+                        subscriptables=[
+                            SubscriptableMapper(None, "tags", None, "tags"),
+                            SubscriptableMapper(None, "contexts", None, "contexts"),
+                        ],
+                    )
                 ),
+            )
+        )
+
+        def selector_func(_query: Query) -> Tuple[str, List[str]]:
+            if random.random() < float(
+                state.get_config("discover_query_percentage", 0)
+            ):
+                return "events", ["discover"]
+
+            return "events", []
+
+        def callback_func(results: List[Tuple[str, QueryResult]]) -> None:
+            _, primary_result = results.pop(0)
+
+            for _, result in results:
+                if result.result["data"] == primary_result.result["data"]:
+                    metrics.increment("query_result", tags={"match": "true"})
+                else:
+                    metrics.increment("query_result", tags={"match": "false"})
+
+        super().__init__(
+            storages=[events_storage, discover_storage],
+            query_pipeline_builder=PipelineDelegator(
```

Comment:
I think I would fit this into the dataset execution pipeline. (just introduced https://github.com/getsentry/snuba/blob/master/snuba/datasets/dataset.py#L55-L79)

So that we can ensure that this is only used for simple queries (until we have a customizable composite pipeline) and that the PipelineDelegator  does not fail when we try to get the planner for an entity. (I think it still says not implemented). I suspect this may be failing now.

Reply:
In order to make this work with the composite queries quickly I think you need to at least provide an implementation here:
https://github.com/getsentry/snuba/blob/master/snuba/pipeline/pipeline_delegator.py#L105 (that returns one planner) so we would skip delegation for composite queries.

For a more stable solution there are a few alternatives:
- the ones we discussed today around being able to define "hints" or some sort of context for the execution pipeline. These would be set by the delegator at the very beginning of the execution pipeline and read at any point (like during storage selection).
- override the query_pipeline_builder for an antity (the discover entity) in the datasetPipelineBuilder so that it can customize how the pipeline builder will work in one of the two branches (https://github.com/getsentry/snuba/blob/master/snuba/datasets/dataset.py#L74). 

(i like the first more as it would work in a generic environment, not only the customization of the execution pipeline, though it would be a longer term solution)

Reply:
Added the implementation that just runs the primary builder and query plan. As for the long term solution, I also like the first option. We can run alternate versions of the query (with different "hints" attached) through the same pipeline rather than require entire alternate pipelines or ways to switch out parts of pipelines to be provided.